pkg/types/contract_reader_cache.go

```diff
@@ -0,0 +1,441 @@
+func (c *CachedContractReader) Unbind(ctx context.Context, bindings []BoundContract) error {
+	return c.base.Bind(ctx, bindings)
+}
+
+func (c *CachedContractReader) Start(ctx context.Context) error {
+	return c.state.StartOnce(CachedContractReaderName, func() error {
+		c.startSubproc(c.clean, cleanInterval)
+		c.startSubproc(c.pollAndCache, pollInterval)
+
+		return nil
+	})
+}
+
+func (c *CachedContractReader) Close() error {
+	return c.state.StopOnce(CachedContractReaderName, func() error {
+		for _, chClose := range c.subprocs {
+			chClose <- struct{}{}
+		}
+
+		return nil
+	})
+}
+
+func (c *CachedContractReader) Ready() error { return nil }
+
+func (c *CachedContractReader) Name() string { return CachedContractReaderName }
+
+func (c *CachedContractReader) HealthReport() map[string]error {
+	return map[string]error{c.Name(): nil}
+}
+
+func (c *CachedContractReader) SetCacheStrategy(readIdentifier string, confidence primitives.ConfidenceLevel, strategy any) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	switch typed := strategy.(type) {
+	case PollAndCache:
+		c.polling = append(c.polling, newCacheKey(readIdentifier, confidence, typed.Params, 0))
+	case TriggerAndCache:
+		c.triggers = append(c.triggers, newCacheKey(readIdentifier, confidence, typed.Params, 0))
+	case FallthroughOnStale, NoCache:
+		// nothing to set up
+	default:
+		return ErrInvalidStrategy
+	}
+
+	c.setStrategy(readIdentifier, strategy)
+
+	return nil
+}
+
+func (c *CachedContractReader) TriggerWithContext(ctx context.Context) {
+	c.mu.RLock()
+	defer c.mu.Unlock()
+
+	for _, key := range c.triggers {
+		if ctx.Err() != nil {
+			break
+		}
+
+		config, ok := c.getStrategy(key.readIdentifier)
+		if !ok {
+			continue
+		}
+
+		triggerConfig := config.(TriggerAndCache)
+
+		_ = c.fallThroughAndCache(ctx, key, triggerConfig.Value)
+	}
+}
+
+func (c *CachedContractReader) getFromCache(key cacheKey, returnVal any) error {
+	c.dataLock.RLock()
+	data, hasData := c.data[key.String()]
+	c.dataLock.RUnlock()
+
+	if !hasData {
+		return ErrNotExist
+	}
+
+	if key.staleness > 0 {
+		// do staleness check
+		if time.Since(data.LastUpdate()) > key.staleness {
+			return ErrNotExist
+		}
+	}
+
+	retVal := reflect.ValueOf(returnVal)
+	dataVal := reflect.ValueOf(data.Value())
+
+	if retVal.Type() != dataVal.Type() {
+		return fmt.Errorf("%w: cached value and return value do not have the same type", ErrTypeMismatch)
+	}
+
+	reflect.Indirect(retVal).Set(reflect.Indirect(dataVal))
+
+	return nil
+}
+
+func (c *CachedContractReader) fallThroughAndCache(ctx context.Context, key cacheKey, returnVal any) error {
+	data := c.getOrCreate(key)
+
+	data.SetLastUpdate(time.Now())
+
+	retBaseType := reflect.Indirect(reflect.New(reflect.Indirect(reflect.ValueOf(returnVal)).Type()))
+	savedVal := reflect.New(retBaseType.Type())
+
+	reflect.Indirect(savedVal).Set(retBaseType)
+
+	cachedValue := savedVal.Interface()
```

Comment:
What is this doing? A copy?

Reply:
Would the overall code be simpler if we called `fallThrough` with `returnVal`, and then copied it to `cachedValue` afterwards? (e.g. only one reflect block afterwards would be necessary)